internal/webservice/routes.go

```diff
@@ -107,7 +110,50 @@ func (ws *WebService) getQueuesPerPartition(w http.ResponseWriter, r *http.Reque
 		errorResponse(w, r, err)
 		return
 	}
-	jsonResponse(w, queues)
+	root, err := buildPartitionQueueTree(r.Context(), queues)
+	if err != nil {
+		errorResponse(w, r, err)
+	}
+	jsonResponse(w, root)
+}
+
+func buildPartitionQueueTree(ctx context.Context, queues []*model.PartitionQueueDAOInfo) (*dao.PartitionQueueDAOInfo, error) {
+	if err := ctx.Err(); err != nil {
+		return nil, err
+	}
+
+	if len(queues) == 0 {
+		return nil, nil
+	}
+
+	queueMap := make(map[string]*dao.PartitionQueueDAOInfo)
+	for _, queue := range queues {
+		queueMap[queue.Id] = &queue.PartitionQueueDAOInfo
+	}
+
+	var rootID string
+	for _, queue := range queues {
+		if !queue.ParentId.Valid {
+			rootID = queue.Id
+			continue
+		}
```

Comment:
What if we have multiple root queues marked as deleted? In `yunikorn`, it's not possible to have more than one active root queue at the same time per partition. However, since we are displaying historical data, we may need to show queues that have been removed as well.

Should the front end adapt to handle a list of root queues instead? Restricting the system to return only one root queue may not be suitable for our use case.

Reply:
Great point!